zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxn.java

```diff
@@ -1274,7 +1295,6 @@ private void cleanAndNotifyState() {
                 eventThread.queueEvent(new WatchedEvent(Event.EventType.None, Event.KeeperState.Disconnected, null));
             }
             clientCnxnSocket.updateNow();
-            clientCnxnSocket.updateLastSendAndHeard();
```

Comment:
Why have you removed this?

Reply:
Because we are not sending nor receiving anything in disconnection. And, session expiration timeout is calculating base on `lastHeard`. If we are going to update it, we will never hit `SessionTimeoutException`.

Besides, `updateLastSendAndHeard` updates both `lastSend` and `lastHeard`. It should be only called for newly established connection. That is one per connection.
1. NIO: https://github.com/apache/zookeeper/blob/bc9afbf8ef1bc6156643d3d05c87fcf8411e9d8f/zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxnSocketNIO.java#L344-L348
2. Netty: https://github.com/apache/zookeeper/blob/bc9afbf8ef1bc6156643d3d05c87fcf8411e9d8f/zookeeper-server/src/main/java/org/apache/zookeeper/ClientCnxnSocketNetty.java#L166-L168